shared/lib-hoppy/hoppy/async_consumer.py

```diff
@@ -0,0 +1,210 @@
+
+    def on_channel_open(self, channel):
+        logging.debug('Consumer - Channel opened')
+        self._channel = channel
+        self._channel.add_on_close_callback(self.on_channel_closed)
+        self.setup_exchange(self.exchange_name)
+
+    def on_channel_closed(self, channel, reason):
+        logging.warning(f'Consumer - Channel {channel} was closed: {reason}')
+        self.close_connection()
+
+    def setup_exchange(self, exchange_name):
+        logging.debug(f'Consumer - Declaring exchange: {self.exchange_name}')
+        self._channel.exchange_declare(exchange=exchange_name,
+                                       exchange_type=self.exchange_type,
+                                       passive=self.passive_declare_exchange,
+                                       durable=self.durable_exchange,
+                                       auto_delete=self.auto_delete_exchange,
+                                       callback=self.on_exchange_declare_ok)
+
+    def on_exchange_declare_ok(self, _unused_frame):
+        logging.debug(f'Consumer - Exchange declared: {self.exchange_name}')
+        self.setup_queue()
+
+    def setup_queue(self):
+        logging.debug(f'Consumer - Declaring queue {self.queue_name}')
+        self._channel.queue_declare(queue=self.queue_name,
+                                    passive=self.passive_declare_queue,
+                                    durable=self.durable_queue,
+                                    exclusive=self.exclusive_queue,
+                                    auto_delete=self.auto_delete_queue,
+                                    callback=self.on_queue_declare_ok)
+
+    def on_queue_declare_ok(self, _unused_frame):
+        logging.debug(f'Consumer - Binding {self.exchange_name} to {self.queue_name} with {self.routing_key}')
+        self._channel.queue_bind(
+            self.queue_name,
+            self.exchange_name,
+            routing_key=self.routing_key,
+            callback=self.on_bind_ok)
+
+    def on_bind_ok(self, _unused_frame):
+        logging.debug(f'Consumer - Queue bound: {self.queue_name}')
+        self._channel.basic_qos(prefetch_count=self._prefetch_count, callback=self.on_basic_qos_ok)
+
+    def on_basic_qos_ok(self, _unused_frame):
+        logging.debug(f'Consumer - QOS set to: {self._prefetch_count}')
+        self.start_consuming()
+
+    def start_consuming(self):
+        logging.debug('Consumer - Issuing consumer related RPC commands')
+        self._channel.add_on_cancel_callback(self.on_consumer_cancelled)
+        self._consumer_tag = self._channel.basic_consume(self.queue_name, self.on_message)
+        self._consuming = True
+
+    def on_consumer_cancelled(self, method_frame):
+        logging.debug(f'Consumer - Consumer was cancelled remotely, shutting down: {method_frame}')
+        if self._channel:
+            self._channel.close()
+
+    def on_message(self, _unused_channel, basic_deliver, properties, body):
```

Comment:
what happens when reply_callback method fails? Do we need another error handler for this?

Reply:
This would kill the connection prompting a reconnect. I'll add in a try:except to catch it and log an error instead of dumping the connection.